auth/dfa.js

```diff
@@ -0,0 +1,348 @@
+      scope = s1;
+      i += 1;
+      j += 1;
+    } else {
+      let z = scopeCompare(s1, s2);
+      if (z < 0) {
+        scope = s1;
+        scopes.push(s1);
+        i += 1;
+      } else {
+        scope = s2;
+        scopes.push(s2);
+        j += 1;
+      }
+    }
+    // If we just added a star scope, we have to skip everything that matches
+    if (scope.endsWith('*')) {
+      let prefix = scope.slice(0, -1);
+      while(i < n && scopes1[i].startsWith(prefix)) {
+        i += 1;
+      }
+      while(j < m && scopes2[j].startsWith(prefix)) {
+        j += 1;
+      }
+    }
+  }
+  while (i < n) {
+    let scope = scopes1[i];
+    scopes.push(scope);
+    i += 1;
+    if (scope.endsWith('*')) {
+      let prefix = scope.slice(0, -1);
+      while(i < n && scopes1[i].startsWith(prefix)) {
+        i += 1;
+      }
+    }
+  }
+  while (j < m) {
+    let scope = scopes2[j];
+    scopes.push(scope);
+    j += 1;
+    if (scope.endsWith('*')) {
+      let prefix = scope.slice(0, -1);
+      while(j < m && scopes2[j].startsWith(prefix)) {
+        j += 1;
+      }
+    }
+  }
+  return scopes;
+};
+
+// Export mergeScopeSets
+exports.mergeScopeSets = mergeScopeSets;
+
+/**
+ * Build a DFA of states on the form:
+ * ```js
+ * {
+ *   'a': // next state if the current character is 'a'
+ *   'end': [] // set of roles granted if the scope ends here
+ *   'prefix': {...} // role granted if the scope matches so far
+ * }
+ * ```
+ *
+ * Must be started from: i = 0, n = roles.length, k = 0
+ */
+let generateDFA = (roles, i, n, k) => {
+  var state = {};
+  var role = roles[i];
+  var current = role.roleId[k];
+  var begin = i;
+  // works because '' and '*' are sorted to top
+  if (current === undefined) {
+    state.end = [role];
+    i += 1;
+    if (i >= n) {
+      return state;
+    }
+    role = roles[i];
+    current = role.roleId[k];
+  }
+  if (current === '*' && role.roleId.length === k + 1) {
+    state.prefix = role;
+    i += 1;
+    if (i >= n) {
+      return state;
+    }
+    role = roles[i];
+    current = role.roleId[k];
+  }
+  var start = i;
+  while(i < n) {
+    role = roles[i];
+    var c = role.roleId[k];
+    if (c !== current) {
+      state[current] = generateDFA(roles, start, i, k + 1);
+      current = c;
+      start = i;
+    }
+    i += 1;
+  }
+  if (start !== i || !state.end) {
+    state[current] = generateDFA(roles, start, i, k + 1);
+  }
+  var star = state['*'];
+  if (!star) {
+    state['*'] = star = {};
+  }
+  if (!state.prefix) {
+    star.end = roles.slice(begin, n);
+  } else {
+    star.end = roles.slice(begin + 1, n);
+    if (state.end) {
+      star.end[0] = roles[begin];
+    }
+  }
+
+  return state;
+};
+
+// Export generateDFA
+exports.generateDFA = generateDFA;
+
+/**
+ * Builds a resolver that given a scope returns a set of roles.
+ *
+ * This also returns a lists of `sets`, where if you replace an entry in the
+ * list of sets it'll be the return value from the resolver.
```

Comment:
This doesn't make much sense.  Lists of sets of.. things? Return value from the resolver given what argument?